### PR TITLE
Add `SporkID` and `ProtocolVersion` to protocol state snapshot

### PR DESCRIPTION
**Pull request by jordanschalm**
This PR adds two global parameters to the protocol state.
* `SporkID` - unique identifier for a particular spork on a particular network
* `ProtocolVersion` - major software version for the current spork

## Changes
* Adds new fields to `protocol.Params`
* Splits `Params` into `InstanceParams` (may be different across different node instances in the same spork) and `GlobalParams` (must be the same across different node instances in the same spork)
* Adds alias to `protocol.Snapshot` to access the state's `GlobalParams` - this is used in bootstrapping to bootstrap the state with these values and maintain the property that snapshots contain sufficient information to bootstrap a state instance

### CODE REVIEW

**Comment by vishalchangrani** [comment]
@jordanschalm  - Should we then be deprecating this [Version](https://github.com/onflow/flow-go/blob/2f7d0dfc4a6b9a072f09a07ca0be631194a15d2b/cmd/build/version.go#L14) ?

**Comment by jordanschalm** [comment]
> Should we then be deprecating this Version ?

I don't think so. That is the specific commit or tag which built a binary, which is useful on its own. The protocol version represents a span of minor versions which are mutually compatible. The protocol version should be a product of the other version (the default value for the flag could be `build.Semver().Major()` but we are still using `0.X.X` and minor version bumps each spork, so I left that out for now.

**Review by vishalchangrani — APPROVED** [review]
lgtm

File: state/protocol/params.go [hunk 1]
```diff
@@ -4,14 +4,44 @@ import (
...
+
+	// SporkID returns the unique identifier for this network within the current spork.
+	// This ID is determined at the beginning of a spork during bootstrapping and is
+	// part of the root protocol state snapshot.
+	SporkID() (flow.Identifier, error)
+
+	// ProtocolVersion returns the protocol version, the major software version
+	// of the protocol software.
+	ProtocolVersion() (uint, error)
+}
```
Comment (yhassanzadeh13) [inline state/protocol/params.go:47]: I think here is a good place to also document the purpose of each of these, for example, what information one may infer from `ChainID`?

File: state/protocol/inmem/encodable.go [hunk 2]
```diff
@@ -51,3 +52,10 @@ type EncodableCluster struct {
 	RootBlock *cluster.Block
 	RootQC    *flow.QuorumCertificate
 }
+
+// EncodableParams is the encoding format for protocol.GlobalParams
```
Comment (yhassanzadeh13) [inline state/protocol/inmem/encodable.go:56]: Not sure the `godoc` correctly reflects the `EncodableParams` purpose.

File: state/protocol/badger/state.go [hunk 3]
```diff
@@ -369,6 +375,34 @@ func (state *State) bootstrapEpoch(root protocol.Snapshot) func(*transaction.Tx)
 	}
 }
 
+// bootstrapSporkInfo bootstraps the protocol state with information about the
+// spork which is used to disambiguate Flow networks.
+func (s *State) bootstrapSporkInfo(root protocol.Snapshot) func(*badger.Txn) error {
```
Comment (yhassanzadeh13) [inline state/protocol/badger/state.go:380]:
```suggestion
func (state *State) bootstrapSporkInfo(root protocol.Snapshot) func(*badger.Txn) error {
```
receiver names are different.

**Review by yhassanzadeh13 — APPROVED** [review]

File: storage/badger/operation/spork.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+// RetrieveSporkID retrieves the spork ID for the present spork.
+func RetrieveSporkID(sporkID *flow.Identifier) func(*badger.Txn) error {
+	return retrieve(makePrefix(codeSporkID), sporkID)
+}
+
+// InsertProtocolVersion inserts the protocol version for the present spork.
+// A single database and protocol state instance spans at most one spork, and
+// a spork has exactly one protocol version for its duration, so this is
+// inserted exactly once, when bootstrapping the state.
+func InsertProtocolVersion(version uint) func(*badger.Txn) error {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply storage/badger/operation/spork.go:25]: The spork ID is for uniquely identifying a spork for purposes like cross-talk prevention. 

The protocol version is for identifying a span of software versions which are mutually compatible for purposes like warning when a binary boots with a protocol state specifying an incompatible version.

File: state/protocol/inmem/encodable.go (hunk 2 above)
Comment (jordanschalm) [reply state/protocol/inmem/encodable.go:56]: `EncodableParams` is just a concrete (`struct` type) representation of `protocol.GlobalParams` (`interface` type), mainly for encoding.

**Pull request merged**